File: Data Structure/DSA/quick_sort.py

lst = [10, 16, 8, 12, 15, 6, 9]


def swap(lst, ind1, ind2):
    temp = lst[ind2]
    lst[ind2] = lst[ind1]
    lst[ind1] = temp


def partition(lst, low, high):
    pivot = low
    i = low+1
    j = high

    while(i < j):
        while(lst[i] <= lst[pivot]):
            i += 1

        while(lst[j] > lst[pivot]):
            j -= 1

        if(i < j):
            swap(lst, i, j)

    swap(lst, j, pivot)
    return j


def quick_sort(lst,  low, high):
    if(low < high):
        sorted_element = partition(lst, low, high)
        quick_sort(lst, low, sorted_element-1)
        quick_sort(lst, sorted_element+1, high)


quick_sort(lst, 0, len(lst)-1)
print(lst)

File: Data Structure/DSA_C/QuickSort.c
#include <stdio.h>

void printArray(int *A, int n)
{
    int i;
    for (i = 0; i <= n - 1; i++)
    {
        printf("%d ", A[i]);
    }
    printf("\n");
}

void swap(int *i, int *j)
{
    int temp = *i;
    *i = *j;
    *j = temp;
}

int partition(int A[], int low, int high)
{
    int pivot = A[low];
    int i = low + 1;
    int j = high;

    do
    {
        while (A[i] <= pivot)
            i++;

        while (A[j] > pivot)
            j--;

        if (i < j)
            swap(&A[i], &A[j]);

    } while (i < j);

    swap(&A[low], &A[j]);
    return j;
}

void QuickSort(int A[], int low, int high)
{
    if (low < high)
    {
        int PartitionIndex = partition(A, low, high);
        QuickSort(A, low, PartitionIndex - 1);
        QuickSort(A, PartitionIndex + 1, high);
    }
}
void main()

{
    int A[] = {99, 98, 91, 97, 96, 95, 94, 93, 92, 95, 94, 91, 90};
    int n = sizeof(A) / sizeof(int);

    printArray(A, n);
    QuickSort(A, 0, n - 1);
    printArray(A, n);
}